src/Compilers/Core/Portable/Compilation/IOperationWalker.cs

```diff
@@ -0,0 +1,394 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.Semantics
+{
+    /// <summary>
+    /// Represents a <see cref="OperationVisitor"/> that descends an entire <see cref="IOperation"/> tree
+    /// visiting each IOperation and its child IOperation nodes in depth-first order.
+    /// </summary>
+    public abstract class OperationWalker : OperationVisitor
+    {
+        private int _recursionDepth;
+
+        private void VisitList<T>(ImmutableArray<T> list) where T : IOperation
+        {
+            if (!list.IsDefault)
+            {
+                for (int i = 0; i < list.Length; i++)
```

Comment:
It's a small detail, but capturing list.Length in a local outside the loop might save one cycle per iteration. I have no idea if our back end compilers are able to recognize that list.Length is a loop invariant.